### DISCUSSION TITLE
OneToOne eager loading

### DISCUSSION THREAD

**Discussion by zgid123** [mikro-orm/mikro-orm#4369] [Q&A] [answered]
Hi,

I have two models

```ts
// A.ts

@Entity()
export class A extends Base {
  @Property()
  title: string;

  @OneToOne({
    entity: () => B,
    mappedBy: 'a',
    ref: true,
  })
  b: Ref<B>;
}

// B.ts
@Entity()
export class B extends Base {
  @Property()
  title: string;

  @OneToOne({
    entity: () => A,
    inversedBy: 'b',
    ref: true,
    eager: false,
  })
  a: Ref<A>;
}
```

When getting A with populate B

```ts
    const a = await this.aRepository.findOne(
      {
        id: 1,
      },
      {
        populate: ['b'],
      },
    );
```

B will be loaded along A, but at this moment, `b.getEntity().a` is loaded too. I guess because `mikro-orm` will assign it from a as reference.

Is there a way to make `a` from `b` is not `initialized`? Because I am using [automapper](https://automapperts.netlify.app/) to do the serializer. When eager loading like that, it will cause an infinite loop.

I am using `6.0.0-dev.69` for my main project. For my test mikro-orm project, it is `6.0.0-dev.15`.

Thanks.

**Comment by B4nan** [comment, marked as answer]
Nope, that's not possible, there is always one instance of the entity in a given context, it's either initialized or not, you can't have one initialized copy and another one that is not. Note that `eager: false` is the default, there is no eager loading in your code snippet.

This is what the ORM serialization is trying to deal with, in latest v6 versions it should give you just the FK in `a.b.a` - try doing `wrap(b).toObject()` to see. But I guess the automapper is kinda a replacement for it? Never used it myself, so not really sure how/for what you are using it.

> Because I am using [automapper](https://automapperts.netlify.app/) to do the serializer. When eager loading like that, it will cause an infinite loop.

Either way, this sounds like a bug on their end to me.

**Reply by zgid123** [reply]
yup, that is a replacement. Lemme try to fix it as `automapper` side. Thanks

**Comment by zgid123** [comment]
Hi @B4nan,

Sorry for bothering you again.

About the association, it seems we always should set the `nullable: true` for the association right?

For example the thing above

```ts
@Entity()
export class A extends Base {
  @Property()
  title: string;

  @OneToOne({
    entity: () => B,
    mappedBy: 'a',
    ref: true,
  })
  b: Ref<B>;
}
```

```ts
@Entity()
export class B extends Base {
  @Property()
  title: string;

  @OneToOne({
    entity: () => A,
    inversedBy: 'b',
    ref: true,
    // nullable: true,
  })
  a: Ref<A>;
}
```

When define like this one, I ran this code

```ts
let a: A = await this.aRepository.findOne(
      {
        id: 1,
      },
      {
        populate: ['b'],
      },
    );

    if (!a) {
      a = this.aRepository.create({
        title: 'test A',
        b: {
          title: 'test B',
        },
      });

      await this.aRepository.getEntityManager().flush();
    } else {
      const b = this.bRepository.create({
        title: 'test aaa',
      });

      wrap(a).assign({
        b,
      });

      wrap(a).assign({
        title: `test a ${new Date().getTime()}`,
      });

      await this.aRepository.getEntityManager().flush();
    }
```

The second run, it will raise error 

```sh
update "bs" set "a_id" = NULL, "updated_at" = '2023-05-21T09:45:49.236Z' where "id" = 1 - null value in column "a_id" of relation "bs" violates not-null constraint
```

I am using `6.0.0-dev.75`.

Don't worry about two wrap assign, because in my main project, there is an condition, so I need to assign twice.

**Reply by B4nan** [reply]
That's weird, no, you can have a non-null relation. I see you work with two different repositories, so could be some context issues - are `this.aRepository.getEntityManager().id` and `this.bRepository.getEntityManager().id` the same? They should. 

You should prefer using EM instead of this madness :] The whole point of the removal of `persist/flush` methods from the repository level was to force people away from this. Repositories are nice if you want to add custom methods, mostly for reading - for writing it's almost always better to just use the EM directly.

**Reply by zgid123** [reply]
here the repository.

https://github.com/zgid123/test-mikro-with-nestjs/tree/test-one-to-one

When I remove the `nullable` for B entity, it will raise error. But set the `nullable: true`, I can create b record without error.

Lemme try the `EM` approach instead

**Reply by zgid123** [reply]
wow, nice. With `EM` approach, it fixes the issue.

Thank you so much.

**Reply by B4nan** [reply]
You really don't want to do this, using global context can lead to (not only) the same issues as this one, and in the long run it will always result in memory leaks (unless you clear the context explicitly). Each request handler needs to have its own context, which can be achieved via middlewares - that is automatically handled by the `@mikro-orm/nestjs` package.

https://github.com/zgid123/test-mikro-with-nestjs/blob/test-one-to-one/workspaces/api-gateway/src/mikro-orm.config.ts#L7

**Reply by zgid123** [reply]
don't worry, this is just for testing. You mentioned this for me once. My main project only use global context when I am using with `gRPC` only (because I still not know how to do the middleware for `gRPC`). When not using `gRPC`, I will use the middleware of mikro-orm.
